src/ui/public/courier/data_source/search_source.js

```diff
@@ -51,280 +51,645 @@
+        request.abort();
+      });
+
+      fetchSoon.these([req]);
+
+      return req.getCompletePromise();
+    }
+
+
+    /**
+     * Fetch all pending requests for this source ASAP
+     * @async
+     */
+    fetchQueued() {
+      return fetchSoon.these(this._myStartableQueued());
+    }
+
+    /**
+     * Cancel all pending requests for this dataSource
+     * @return {undefined}
+     */
+    cancelQueued() {
+      requestQueue
+        .filter(req => req.source === this)
+        .forEach(req => req.abort());
+    }
+
+    /**
+     * Completely destroy the SearchSource.
+     * @return {undefined}
+     */
+    destroy() {
+      this.cancelQueued();
+      this._requestStartHandlers.length = 0;
     }
-  };
-
-  SearchSource.prototype._filterPredicates = [
-    (filter) => {
-      // remove null/undefined filters
-      return filter;
-    },
-    (filter) => {
-      const disabled = _.get(filter, 'meta.disabled');
-      return disabled === undefined || disabled === false;
-    },
-    (filter, state) => {
-      if (!config.get('courier:ignoreFilterIfFieldNotInIndex')) {
-        return true;
+
+    /**
+     *  Add a handler that will be notified whenever requests start
+     *  @param  {Function} handler
+     *  @return {undefined}
+     */
+    onRequestStart(handler) {
+      this._requestStartHandlers.push(handler);
+    }
+
+    /**
+     *  Called by requests of this search source when they are started
+     *  @param  {Courier.Request} request
+     *  @return {Promise<undefined>}
+     */
+    requestIsStarting(request) {
+      this.activeFetchCount = (this.activeFetchCount || 0) + 1;
+      this.history = [request];
+
+      return Promise
+        .map(this._requestStartHandlers, fn => fn(this, request))
+        .then(_.noop);
+    }
+
+
+    /******
+     * PRIVATE APIS
+     ******/
+
+    _myStartableQueued() {
+      return requestQueue
+        .getStartable()
+        .filter(req => req.source === this);
+    }
+
+    /**
+     * Gets the type of the DataSource
+     * @return {string}
+     */
+    _getType() {
+      return 'search';
+    }
+
+    /**
+     * Create a common search request object, which should
+     * be put into the pending request queye, for this search
+     * source
+     *
+     * @param {Deferred} defer - the deferred object that should be resolved
+     *                         when the request is complete
+     * @return {SearchRequest}
+     */
+    _createRequest(defer) {
+      return new SearchRequest(this, defer);
+    }
+
+    /**
+     * Used to merge properties into the state within ._flatten().
+     * The state is passed in and modified by the function
+     *
+     * @param  {object} state - the current merged state
+     * @param  {*} val - the value at `key`
+     * @param  {*} key - The key of `val`
+     * @return {undefined}
+     */
+    _mergeProp(state, val, key) {
+      if (typeof val === 'function') {
+        const source = this;
```

Comment:
Stopping the review here, as I realize you're just moving code around for the purpose of refactoring, and this code is already in the codebase. That said, I'd flag this as something that would benefit additional refactoring at some point. It's fairly difficult to follow. At ~700 loc, this feels like a "god class". Not sure what to do about it or whether it'll ever be worth it, but a lot of this appears as if it'd benefit from a conversion from OOP to a more functional approach, as it's conflating data transforms with a lot of other things (mutating the DOM, etc), so separating those things would be nice.

Reply:
Agree completely!  Something I've been wanting to do forever is clean this code up but it's tricky and I've been taking very small steps to ensure forward progress (if I try to tackle too much at once I get down the rabbit hole and the changes become too huge and difficult to find the reason for bugs).